### PR TITLE
Added possibility to download only some files

### PR DESCRIPTION
**Pull request by maoueh**
Added a `pitreos files` command at the same time to list available files in
a given backup.

### CODE REVIEW

File: lib/files.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+	bm, err := p.downloadBackupIndex(backupName)
+	if err != nil {
+		return err
+	}
+
+	if bm.Version != p.filemetaVersion {
+		return fmt.Errorf("Incompatible version of backupIndex. Expected: %s, found: %s.", p.filemetaVersion, bm.Version)
+	}
+
+	filterRegex, err := regexp.Compile(filter)
```
Comment (maoueh) [inline lib/files.go]: Or?

```suggestion
	filterRegex, err := regexp.CompilePOSIX(filter)
```

Comment (sduchesneau) [reply lib/files.go]: perl, perl, perl !!

File: lib/files.go [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	if err != nil {
+		return err
+	}
+
+	w := new(tabwriter.Writer)
+	w.Init(os.Stdout, 5, 5, 1, '\t', tabwriter.AlignRight)
+
+	for _, file := range bm.Files {
+		if !filterRegex.MatchString(file.FileName) {
```
Comment (maoueh) [inline lib/files.go]: This logic is now duplicated between restore and files. I should probably refactor to have a single code path. I though about something like:

(Pseudo-code)
```
forEachBackupFiles ...
  do something
```

Question for this one is `forEachBackupFiles(backupName` or `forEachBackupFiles(backupObject)`. Former version would perform the download of index. Latter version would receive the backup object directly.

Thoughts?

**Comment by maoueh** [comment]
Example:

```
$ pitreos files -s gs://eoscanada-mainnet-nodeos-backup/pitreos 2018-10-19-10-00-08--default "state|reversible"
2018/10/19 14:38:15 Not using config file: Config File ".pitreos" Not Found in "[/Users/maoueh/go/src/github.com/eoscanada/pitreos /Users/maoueh]"
2018/10/19 14:38:15 Using storage: gs://eoscanada-mainnet-nodeos-backup/pitreos
356 MB	blocks/reversible/shared_memory.bin
2.9 kB	blocks/reversible/shared_memory.meta
3.7 MB	state/forkdb.dat
67 GB	state/shared_memory.bin
2.9 kB	state/shared_memory.meta
```

File: lib/files.go (hunk 2 above)
Comment (sduchesneau) [reply lib/files.go]: you want to go into functional programming for that simple regexp filter ?
not a big fan considering the context...
maybe:
``` 
	for _, file := range filtered(bm.Files, filter) { ... }
```
where filtered is just a wrapper function over "regexp.compile" et "regexp.matchString"...
but then again, you should handle the case where the filter is not a valid regex...

so you end up not wrapping anything useful...

Comment (sduchesneau) [reply lib/files.go]: wantedFiles, err := applyFilter(bm, filter)
errorCheck("applying requested regex filter", err)

...
for _, file := wantedFiles

File: lib/types.go [hunk 3]
```diff
@@ -31,3 +32,48 @@ type ListableBackup struct {
...
+			continue
+		}
+
+		estimatedDiskSize += uint64(chunk.End - chunk.Start)
+	}
+
+	return estimatedDiskSize, nil
+}
+
+func (backupIndex *BackupIndex) FindFilesMatching(filter string) ([]*FileIndex, error) {
```
Comment (maoueh) [inline lib/types.go]: Went with this method to remove duplication.

**Comment by maoueh** [comment]
Resolved all comments, merging. Feel free to review again and comment if needed, I would make change in an upcoming PR.

**Pull request merged**